File: flashml-cloud/apps/api/flashml_cloud_api/contributions.py
"""What one account has contributed — a counter, and deliberately not a wallet.

``public.contributions`` has recorded accepted work per (machine, job, task)
since the first migration, and until this module existed a person could not
see their own total anywhere. Credit surfaced only inside one job's panel —
"which machine did the work on THIS job" — so somebody who had donated forty
hours across five jobs had no way to learn it. That is the gap this fills,
and the barter premise of the product does not survive without it.

**Nothing in this system debits these numbers.** There is no spend path, no
balance, no ceiling and no expiry: every credit ever written is still there
and always will be. So the shape below counts *tasks* and *jobs* and says so
in its field names. It must never grow a `balance`, a `remaining`, or a
`credits` — a name that implies a drawdown promises an exchange rate this
product has not designed and cannot honour, and a person who believes they
have a balance is owed one.

The arithmetic lives here as pure functions over rows a query supplied, for
the same reason ``storage.py`` keeps the budget rule out of ``db`` and
``metrics.py`` keeps the goodput rule out of it: the rule is a policy
statement that must be readable and testable without a database, and the
counts are facts. A policy that lives inside a SQL string cannot be read or
argued with, and the ordering, the summing and the null-handling below are
all policy.

**What this counts, and what it refuses to count.** The measurement in ``db``
reads ``contributions``, the credit ledger, and never ``attempts``. Those are
different things with different guarantees:

- ``contributions`` is accepted work. Its unique index (migration 0003) makes
  a duplicate credit impossible at the schema level, so the total here cannot
  drift upward under a retried round callback or a restarted driver.
- ``attempts`` is the lease → (job, task) mapping the completion proxy needs
  because the complete hop carries no job id. It is written by ONE of the two
  credit paths. ``fedavg.on_round`` credits a federated round's contributors
  straight into the ledger and writes no attempts row at all — which is why
  production on 2026-08-03 read 26 credits against 16 attempts on one machine,
  the ten-row gap being federated work. Counting leases would have silently
  dropped every one of them, and it would also have counted a claim that was
  never accepted, which is hard rule 4 exactly backwards.
"""

from __future__ import annotations

from collections.abc import Iterable, Mapping
from datetime import datetime, timezone
from typing import Any


def report(
    machines: Iterable[Mapping[str, Any]], jobs_contributed_to: int
) -> dict[str, Any]:
    """Assemble the response body from the rows a query supplied.

    Every key is always present, including for an account that has
    contributed nothing: **zero is a real answer**, not a 404 and not an
    error. Somebody who just signed up has contributed nothing, that is the
    single most common state this panel renders, and a field that disappears
    when it has nothing to say arrives at the console as ``undefined`` and
    draws as a blank tile that looks like a loading state.

    ``accepted_tasks`` is DERIVED from the machine rows rather than counted a
    second time in SQL. Two independent counts of the same thing eventually
    disagree, and a headline number that does not add up to the list beneath
    it reads as a bug in whichever half is actually right. ``machines`` is
    scoped to one owner and every contribution row belongs to exactly one
    machine, so the sum is not an approximation — it is the same count.

    ``jobs_contributed_to`` cannot be derived the same way: two machines can
    work the same job, and summing per-machine job counts would report it
    twice. It arrives as a distinct count from the query and is passed
    through.
    """
    rows = [_machine(row) for row in machines]

    # Most-contributed first, with the id as tiebreak. The tiebreak is not
    # decoration: equal counts are the ordinary case on a two-laptop fleet,
    # and without it the panel reorders itself between polls, which looks
    # like work moving between machines and is not. Sorted here rather than
    # in the ORDER BY so the guarantee can be tested without a database.
    rows.sort(key=lambda row: (-row["accepted_tasks"], row["machine_id"]))

    return {
        "accepted_tasks": sum(row["accepted_tasks"] for row in rows),
        "jobs_contributed_to": int(jobs_contributed_to),
        "machines": rows,
    }


def _machine(row: Mapping[str, Any]) -> dict[str, Any]:
    """One machine's entry, with both of its nullable fields left null.

    ``hostname`` and ``last_seen_at`` are genuinely absent for real machines:
    an agent supplies its hostname at enrolment and may not have one, and
    ``last_seen_at`` stays null until the first heartbeat. A placeholder in
    either is worse than the null — "Unknown machine" is indistinguishable
    from a host actually called that, and a substituted ``now()`` renders as
    a moment the machine was never seen at.

    ``machine_id`` is stringified because psycopg hands back ``uuid.UUID``,
    which the JSON encoder cannot serialise: left alone, this route would 500
    for the first account that ever contributed and work perfectly for
    everyone who has not.
    """
    return {
        "machine_id": str(row["machine_id"]),
        "hostname": row.get("hostname"),
        "accepted_tasks": int(row["accepted_tasks"]),
        "last_seen_at": _instant(row.get("last_seen_at")),
    }


def _instant(value: Any) -> str | None:
    """A timestamp as unambiguous ISO-8601 UTC, or None.

    ``_jsonable`` in ``app`` renders datetimes with ``str()``, which gives
    ``2026-08-08 00:00:00+00:00`` — a space instead of a ``T``, which Safari's
    ``new Date()`` rejects outright. The older machine routes are stuck with
    that spelling because consumers already parse it; this route is new, so it
    can serve the form every runtime agrees on.

    Converted to UTC, never relabelled: psycopg returns the value in the
    database session's timezone, and stamping a ``Z`` on an offset that is not
    zero is an error of hours that looks exactly like a correct answer.
    """
    if not isinstance(value, datetime):
        return None
    return (
        value.astimezone(timezone.utc)
        .replace(tzinfo=None)
        .isoformat(timespec="seconds")
        + "Z"
    )

File: flashml-cloud/apps/web/lib/contributions.ts
/** Turning `GET /v1alpha1/me/contributions` into something the account page
 * can render. Same reasoning as every other module in this family — see
 * `lib/job-result.ts`'s header for why the logic lives here and the
 * component around it stays markup.
 *
 * FlashML's premise is that people contribute machines and get credited for
 * it, and until this route existed that credit was visible only inside one
 * job's own page. This module carries the one rule that matters once the
 * total is account-wide: these are COUNTERS, not a currency. Nothing here is
 * ever spent, redeemed, or drawn down — a machine only ever adds to
 * `accepted_tasks`, never subtracts from it — so every string this module
 * produces is a running tally, past tense, with no notion of a balance to
 * spend against. "41 tasks accepted" is honest; "41 credits available"
 * claims something this system does not do.
 *
 * `hostname` and `last_seen_at` are genuinely nullable per machine: an agent
 * that has never reported a hostname, or a machine that has never been
 * leased a task at all, is a real and different state from "we know and
 * it's blank". Both get their own explicit label below — never an empty
 * string standing in for "unknown", never a fabricated date standing in for
 * "never happened".
 *
 * Since machines became deletable there is a third state behind a null
 * hostname: a machine whose owner retired it, whose credit is deliberately
 * still counted here. `hostnameLabel` below is where those three are told
 * apart, and how.
 */

import type { ContributionMachine, MyContributions } from "./cloud-api";
import { relativeTime } from "./machine-status";

export interface ContributionMachineRow {
  machineId: string;
  /** The machine's own reported hostname, or an explicit label saying why
   * there is not one. */
  hostnameLabel: string;
  /** True only when `hostnameLabel` is the machine's own hostname. The two
   * substitutes below are sentences about a machine, not names of one, so
   * the panel must not set them in the mono face it uses for ids — a label
   * that looks like a hostname is indistinguishable from a host actually
   * called that. */
  hostnameKnown: boolean;
  acceptedTasks: number;
  /** `relativeTime` already renders a null `last_seen_at` as `"never"` —
   * reused rather than reimplemented so a machine's last-seen time reads
   * identically here and on `account/machines`, the other place this exact
   * field is shown. */
  lastSeenLabel: string;
}

export interface ContributionsSummary {
  acceptedTasks: number;
  jobsContributedTo: number;
  machines: ContributionMachineRow[];
}

const NO_HOSTNAME_LABEL = "No hostname reported";
const DELETED_LABEL = "Deleted machine";

/**
 * What to call a machine with no hostname — and there are now two reasons
 * for that, which are not the same fact.
 *
 * Deleting a machine tombstones its row and nulls every column that
 * described the device, `name` and `last_seen_at` in the same UPDATE, while
 * deliberately keeping its accepted-work credit: a contribution total that
 * FELL because somebody tidied their fleet would be indistinguishable from a
 * bug. So this list can now contain a machine that no longer exists, and
 * calling it "No hostname reported" describes an agent that never introduced
 * itself — which is a different, and still real, machine.
 *
 * BOTH device fields null is the signal, not either alone. A machine only
 * appears in this payload because it was credited for accepted work, and
 * work arrives over a heartbeat, so a credited machine has a `last_seen_at`
 * — unless something scrubbed it, and the delete is the only thing that
 * does. The residual false positive is a machine that contributed, reported
 * no hostname and had its last-seen cleared some other way; there is no such
 * path today, and the row still says something true of a machine rather than
 * nothing at all.
 *
 * The alternative — labelling every null hostname "Deleted machine" — reads
 * as certainty this payload does not carry, and would tell somebody a laptop
 * that is enrolled and running right now has been deleted.
 */
function hostnameLabel(machine: ContributionMachine): string {
  if (machine.hostname) return machine.hostname;
  return machine.last_seen_at === null ? DELETED_LABEL : NO_HOSTNAME_LABEL;
}

export function summariseContributions(
  payload: MyContributions
): ContributionsSummary {
  return {
    acceptedTasks: payload.accepted_tasks,
    jobsContributedTo: payload.jobs_contributed_to,
    machines: payload.machines.map((m) => ({
      machineId: m.machine_id,
      hostnameLabel: hostnameLabel(m),
      hostnameKnown: Boolean(m.hostname),
      acceptedTasks: m.accepted_tasks,
      lastSeenLabel: relativeTime(m.last_seen_at),
    })),
  };
}
